sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ConsumeValueTaskCorrectly.cs

```diff
@@ -0,0 +1,225 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2019 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ConsumeValueTaskCorrectly : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S5034";
+
+        // 'await', 'AsTask', 'Result' and '.GetAwaiter().GetResult()' should be called only once on a ValueTask
+        private const string ConsumeOnlyOnceMessage = "Refactor this 'ValueTask' usage to consume it only once.";
+
+        private static readonly DiagnosticDescriptor ConsumeOnlyOnce =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, ConsumeOnlyOnceMessage, RspecStrings.ResourceManager);
+
+        // This should be called only when 'readTask.IsCompletedSuccessfully' is not called before
```

Comment:
I don't like the statement `is not called before` in the above comment.
It makes me believe that the following case is supported:
```cs
if (!valueTask.IsCompletedSuccessfully)
{
  return weNeedToAwait(valueTask);
}
return valueTask.Result;
```
Is the above case supported or is it a FP?
I think statements used in other comments better represent the logic `called inside an 'if (valueTask.IsCompletedSuccessfully)'`